c10/util/intrusive_ptr.h

```diff
@@ -198,12 +219,23 @@ TTarget* assign_ptr_(TTarget* rhs) {
 
 // The only requirement for refcount increment is that it happens-before
 // decrement, so no additional memory ordering is needed.
-inline uint32_t atomic_refcount_increment(std::atomic<uint32_t>& refcount) {
-  return refcount.fetch_add(1, std::memory_order_relaxed) + 1;
+inline uint64_t atomic_combined_refcount_increment(
+    std::atomic<uint64_t>& combined_refcount,
+    uint64_t inc) {
+  return combined_refcount.fetch_add(inc, std::memory_order_relaxed) + inc;
 }
 
-inline uint32_t atomic_weakcount_increment(std::atomic<uint32_t>& weakcount) {
-  return weakcount.fetch_add(1, std::memory_order_relaxed) + 1;
+inline uint32_t atomic_refcount_increment(
+    std::atomic<uint64_t>& combined_refcount) {
+  return (uint32_t)(atomic_combined_refcount_increment(
+      combined_refcount, kReferenceCountOne));
+}
+
+inline uint32_t atomic_weakcount_increment(
+    std::atomic<uint64_t>& combined_refcount) {
+  return (uint32_t)(atomic_combined_refcount_increment(
+                        combined_refcount, kWeakReferenceCountOne) >>
+                    32);
```

Comment:
The 32 offset is used a lot as a magic number. Better to make it a constexpr variable later in case we want to change the offset at a later time, improve code as documentation about why?

Reply:
> The 32 offset is used a lot as a magic number.

personally I think it's fine in this case because we are using uint32_t and uint64_t,so the sources for the numbers are right there

Reply:
Yeah, wouldn't normally flag it here, just thinking ahead in case we need to support exotic architectures at some point because it's such an important part of the code

Reply:
Introduced utility functions refcount/weakcount in namespace::detail to abstract away how refcount and weakcount are extracted from the combined_refcount